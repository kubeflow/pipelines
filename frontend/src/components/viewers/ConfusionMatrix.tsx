/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Viewer, { ViewerConfig, PlotType } from './Viewer';
import { color, commonCss, fontsize } from '../../Css';
import { classes, stylesheet } from 'typestyle';

const legendNotches = 5;

export interface ConfusionMatrixConfig extends ViewerConfig {
  data: number[][];
  axes: string[];
  labels: string[];
  type: PlotType;
}

interface ConfusionMatrixProps {
  configs: ConfusionMatrixConfig[];
  maxDimension?: number;
}

interface ConfusionMatrixState {
  activeCell: [number, number];
}

class ConfusionMatrix extends Viewer<ConfusionMatrixProps, ConfusionMatrixState> {
  private _opacities: number[][] = [];
  private _config = this.props.configs[0];
  private _max =
    this._config &&
    Math.max(...this._config.data.map(d => d.map(n => +n)).map(d => Math.max(...d)));
  private _minRegularCellDimension = 15;
  private _maxRegularCellDimension = 80;
  private _cellDimension = this._config
    ? Math.max(
        Math.min(
          (this.props.maxDimension || 700) / this._config.data.length,
          this._maxRegularCellDimension,
        ),
        this._minRegularCellDimension,
      ) - 1
    : 0;
  private _shrinkThreshold = 600;

  private _css = stylesheet({
    activeLabel: {
      borderRadius: 5,
      color: color.theme,
      fontWeight: 'bold',
    },
    cell: {
      border: 'solid 1px ' + color.background,
      fontSize: this._isSmall() ? fontsize.small : fontsize.base,
      height: this._cellDimension,
      minHeight: this._cellDimension,
      minWidth: this._cellDimension,
      position: 'relative',
      textAlign: 'center',
      verticalAlign: 'middle',
      width: this._cellDimension,
    },
    legend: {
      background: `linear-gradient(${color.theme}, ${color.background})`,
      borderRight: 'solid 1px #777',
      marginLeft: 20,
      minWidth: 10,
      position: 'relative',
      width: 10,
    },
    legendLabel: {
      left: 15,
      position: 'absolute',
      top: -7,
    },
    legendNotch: {
      borderTop: 'solid 1px #777',
      left: '100%',
      paddingLeft: 5,
      position: 'absolute',
      width: 5,
    },
    overlay: {
      backgroundColor: '#000',
      bottom: 0,
      left: 0,
      opacity: 0,
      position: 'absolute',
      right: 0,
      top: 0,
    },
    root: {
      flexGrow: 1,
      justifyContent: 'center',
      margin: 'auto',
      pointerEvents: this._isSmall() ? 'none' : 'initial', // Disable interaction for snapshot view
      position: 'relative',
      width: 'fit-content',
    },
    xAxisLabel: {
      color: color.foreground,
      fontSize: 15,
      fontWeight: 'bold',
      paddingLeft: 20,
      position: 'absolute',
    },
    xlabel: {
      marginLeft: 15,
      overflow: 'hidden',
      position: 'absolute',
      textAlign: 'left',
      textOverflow: 'ellipsis',
      transform: 'rotate(60deg)',
      transformOrigin: 'left',
      whiteSpace: 'nowrap',
      width: 150,
    },
    yAxisLabel: {
      color: color.foreground,
      fontSize: 15,
      height: 25,
      paddingRight: 20,
      textAlign: 'right',
    },
    ylabel: {
      lineHeight: `${this._cellDimension}px`,
      marginRight: 10,
      minWidth: this._cellDimension,
      textAlign: 'right',
      whiteSpace: 'nowrap',
    },
  });

  constructor(props: any) {
    super(props);

    if (!this._config) {
      return;
    }

    this.state = {
      activeCell: [-1, -1],
    };

    for (const i of this._config.data) {
      const row = [];
      for (const j of i) {
        row.push(+j / this._max);
      }
      this._opacities.push(row);
    }
  }

  public getDisplayName(): string {
    return 'Confusion matrix';
  }

  public render(): JSX.Element | null {
    if (!this._config) {
      return null;
    }

    const [activeRow, activeCol] = this.state.activeCell;
    const [xAxisLabel, yAxisLabel] = this._config.axes;
    const small = this._isSmall();

    return (
      <div className={classes(commonCss.flex, this._css.root)}>
        <table>
          <tbody>
            {!small && (
              <tr>
                <td className={this._css.yAxisLabel}>{yAxisLabel}</td>
              </tr>
            )}
            {this._config.data.map((row, r) => (
              <tr key={r}>
                {!small && (
                  <td>
                    <div
                      className={classes(
                        this._css.ylabel,
                        r === activeRow ? this._css.activeLabel : '',
                      )}
                    >
                      {this._config.labels[r]}
                    </div>
                  </td>
                )}
                {row.map((cell, c) => (
                  <td
                    key={c}
                    className={this._css.cell}
                    style={{
                      backgroundColor: `rgba(41, 121, 255, ${this._opacities[r][c]})`,
                      color: this._opacities[r][c] < 0.6 ? color.foreground : color.background,
                    }}
                    onMouseOver={() => this.setState({ activeCell: [r, c] })}
                  >
                    <div
                      className={this._css.overlay}
                      style={{
                        opacity: r === activeRow || c === activeCol ? 0.05 : 0,
                      }}
                    />
                    {cell}
                  </td>
                ))}
              </tr>
            ))}

            {/* Footer */}
            {!small && (
              <tr>
                <th className={this._css.xlabel} />
                {this._config.labels.map((label, i) => (
                  <th key={i}>
                    <div
                      className={classes(
                        i === activeCol ? this._css.activeLabel : '',
                        this._css.xlabel,
                      )}
                    >
                      {label}
                    </div>
                  </th>
                ))}
                <td className={this._css.xAxisLabel}>{xAxisLabel}</td>
              </tr>
            )}
          </tbody>
        </table>

        {!small && (
          <div
            className={this._css.legend}
            style={{ height: 0.75 * this._config.data.length * this._cellDimension }}
          >
            <div className={this._css.legendNotch} style={{ top: 0 }}>
              <span className={this._css.legendLabel}>{this._max}</span>
            </div>
            {new Array(legendNotches).fill(0).map((_, i) => (
              <div
                key={i}
                className={this._css.legendNotch}
                style={{ top: ((legendNotches - i) / legendNotches) * 100 + '%' }}
              >
                <span className={this._css.legendLabel}>
                  {Math.floor((i / legendNotches) * this._max)}
                </span>
              </div>
            ))}
          </div>
        )}
      </div>
    );
  }

  private _isSmall(): boolean {
    return !!this.props.maxDimension && this.props.maxDimension < this._shrinkThreshold;
  }
}

export default ConfusionMatrix;
