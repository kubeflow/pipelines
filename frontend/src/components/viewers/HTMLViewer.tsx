/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Viewer, { ViewerConfig } from './Viewer';
import { color } from '../../Css';
import { stylesheet } from 'typestyle';

export interface HTMLViewerConfig extends ViewerConfig {
  htmlContent: string;
}

interface HTMLViewerProps {
  configs: HTMLViewerConfig[];
  maxDimension?: number;
}

class HTMLViewer extends Viewer<HTMLViewerProps, any> {
  private _iframeRef = React.createRef<HTMLIFrameElement>();
  private _config = this.props.configs[0];

  private _css = stylesheet({
    iframe: {
      border: '1px solid ' + color.divider,
      boxSizing: 'border-box',
      flexGrow: 1,
      height: this.props.maxDimension ? this.props.maxDimension : 'initial',
      minHeight: this.props.maxDimension ? this.props.maxDimension : 600,
      width: this.props.maxDimension ? this.props.maxDimension : '100%',
    },
  });

  public getDisplayName(): string {
    return 'Static HTML';
  }

  public componentDidMount(): void {
    // TODO: iframe.srcdoc doesn't work on Edge yet. It's been added, but not
    // yet rolled out as of the time of writing this (6/14/18):
    // https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/12375527/
    // I'm using this since it seems like the safest way to insert HTML into an
    // iframe, while allowing Javascript, but without needing to require
    // "allow-same-origin" sandbox rule.
    if (this._iframeRef.current) {
      this._iframeRef.current!.srcdoc = this._config.htmlContent;
    }
  }

  public render(): JSX.Element | null {
    if (!this._config) {
      return null;
    }

    return (
      // TODO: fix this
      // eslint-disable-next-line jsx-a11y/iframe-has-title
      <iframe
        ref={this._iframeRef}
        src='about:blank'
        className={this._css.iframe}
        sandbox='allow-scripts'
      />
    );
  }
}

export default HTMLViewer;
