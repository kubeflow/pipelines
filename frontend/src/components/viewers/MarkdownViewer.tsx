/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Viewer, { ViewerConfig } from './Viewer';
import { cssRaw } from 'typestyle';
import Markdown from 'markdown-to-jsx';

cssRaw(`
.markdown-viewer h1,
.markdown-viewer h2,
.markdown-viewer h3,
.markdown-viewer h4,
.markdown-viewer h5,
.markdown-viewer h6 {
position: relative;
  margin-top: 1em;
  margin-bottom: 16px;
  font-size: initial;
  font-weight: 700;
  line-height: 1.4;
}

.markdown-viewer h1,
.markdown-viewer h2 {
  padding-bottom: .3em;
  border-bottom: 1px solid #eee;
}

.markdown-viewer code,
.markdown-viewer pre {
  background-color: #f8f8f8;
  border-radius: 5px;
}

.markdown-viewer pre {
  border: solid 1px #eee;
  margin: 7px 0;
  padding: 7px;
}
`);

export interface MarkdownViewerConfig extends ViewerConfig {
  markdownContent: string;
}

interface MarkdownViewerProps {
  configs: MarkdownViewerConfig[];
  maxDimension?: number;
}

class MarkdownViewer extends Viewer<MarkdownViewerProps, any> {
  private _config = this.props.configs[0];

  public getDisplayName(): string {
    return 'Markdown';
  }

  public render(): JSX.Element | null {
    if (!this._config) {
      return null;
    }
    return (
      <div className='markdown-viewer'>
        <Markdown>{this._config.markdownContent}</Markdown>
      </div>
    );
  }
}

export default MarkdownViewer;
