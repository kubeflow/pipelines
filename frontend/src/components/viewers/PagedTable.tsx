/*
 * Copyright 2018 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TablePagination from '@material-ui/core/TablePagination';
import TableRow from '@material-ui/core/TableRow';
import TableSortLabel from '@material-ui/core/TableSortLabel';
import Tooltip from '@material-ui/core/Tooltip';
import Viewer, { ViewerConfig, PlotType } from './Viewer';
import { color, fontsize, commonCss } from '../../Css';
import { stylesheet } from 'typestyle';

enum SortOrder {
  ASC = 'asc',
  DESC = 'desc',
}

export interface PagedTableConfig extends ViewerConfig {
  data: string[][];
  labels: string[];
  type: PlotType;
}

interface PagedTableProps {
  configs: PagedTableConfig[];
  maxDimension?: number;
}

interface PagedTableState {
  order: SortOrder;
  orderBy: number;
  page: number;
  rowsPerPage: number;
}

class PagedTable extends Viewer<PagedTableProps, PagedTableState> {
  private _shrinkThreshold = 600;
  private _config = this.props.configs[0];
  private _rowHeight = 30;

  private _css = stylesheet({
    cell: {
      borderRight: 'solid 1px ' + color.divider,
      color: color.foreground,
      fontSize: this._isSmall() ? fontsize.small : fontsize.base,
      paddingLeft: this._isSmall() ? 5 : 'invalid',
      paddingRight: 5,
      pointerEvents: this._isSmall() ? 'none' : 'initial',
    },
    columnName: {
      fontSize: this._isSmall() ? fontsize.base : fontsize.medium,
      fontWeight: 'bold',
      paddingLeft: this._isSmall() ? 5 : 'invalid',
    },
    row: {
      borderBottom: '1px solid #ddd',
      height: this._isSmall() ? 25 : this._rowHeight,
    },
  });

  constructor(props: any) {
    super(props);
    this.state = {
      order: SortOrder.ASC,
      orderBy: 0,
      page: 0,
      rowsPerPage: 10,
    };
  }

  public getDisplayName(): string {
    return 'Table';
  }

  public render(): JSX.Element | null {
    if (!this._config) {
      return null;
    }

    const { data, labels } = this._config;
    const { order, orderBy, rowsPerPage, page } = this.state;
    const emptyRows = rowsPerPage - Math.min(rowsPerPage, data.length - page * rowsPerPage);

    return (
      <div style={{ width: '100%' }} className={commonCss.page}>
        <Table style={{ display: 'block', overflow: 'auto' }}>
          <TableHead>
            <TableRow>
              {labels.map((label, i) => {
                return (
                  <TableCell
                    className={this._css.columnName}
                    key={i}
                    sortDirection={orderBy === i ? order : false}
                  >
                    <Tooltip title='Sort' enterDelay={300}>
                      <TableSortLabel
                        active={orderBy === i}
                        direction={order}
                        onClick={this._handleSort(i)}
                      >
                        {label}
                      </TableSortLabel>
                    </Tooltip>
                  </TableCell>
                );
              }, this)}
            </TableRow>
          </TableHead>

          <TableBody>
            {this._stableSort(data)
              .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
              .map((row, index) => {
                return (
                  <TableRow hover={true} tabIndex={-1} key={index} className={this._css.row}>
                    {row.map((cell, i) => (
                      <TableCell key={i} className={this._css.cell}>
                        {cell}
                      </TableCell>
                    ))}
                  </TableRow>
                );
              })}
            {emptyRows > 0 && (
              <TableRow style={{ height: this._rowHeight * emptyRows }}>
                <TableCell colSpan={6} />
              </TableRow>
            )}
          </TableBody>
        </Table>

        <TablePagination
          component='div'
          count={data.length}
          rowsPerPage={rowsPerPage}
          page={page}
          onChangePage={this._handleChangePage}
          onChangeRowsPerPage={this._handleChangeRowsPerPage}
        />
      </div>
    );
  }

  private _handleSort = (index: number) => () => {
    const orderBy = index;
    let order = SortOrder.ASC;

    if (this.state.orderBy === index && this.state.order === SortOrder.ASC) {
      order = SortOrder.DESC;
    }

    this.setState({ order, orderBy });
  };

  private _handleChangePage = (event: any, page: number) => {
    this.setState({ page });
  };

  private _handleChangeRowsPerPage = (event: any) => {
    this.setState({ rowsPerPage: event.target.value });
  };

  private _isSmall(): boolean {
    return !!this.props.maxDimension && this.props.maxDimension < this._shrinkThreshold;
  }

  private _stableSort(array: string[][]): string[][] {
    const stabilizedThis = array.map((row: string[], index: number): [string[], number] => [
      row,
      index,
    ]);

    const compareFn = this._getSorting(this.state.order, this.state.orderBy);

    stabilizedThis.sort((a: [string[], number], b: [string[], number]) => {
      const order = compareFn(a[0], b[0]);
      if (order !== 0) {
        return order;
      }
      return a[1] - b[1];
    });
    return stabilizedThis.map((el: [string[], number]) => el[0]);
  }

  private _desc(a: string[], b: string[], orderBy: number): number {
    if (b[orderBy] < a[orderBy]) {
      return -1;
    }
    if (b[orderBy] > a[orderBy]) {
      return 1;
    }
    return 0;
  }

  private _getSorting(order: SortOrder, orderBy: number): (a: any, b: any) => number {
    return order === SortOrder.DESC
      ? (a: any, b: any) => this._desc(a, b, orderBy)
      : (a: any, b: any) => -this._desc(a, b, orderBy);
  }
}

export default PagedTable;
