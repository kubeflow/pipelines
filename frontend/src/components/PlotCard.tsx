/*
 * Copyright 2018 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Button from '@material-ui/core/Button';
import CloseIcon from '@material-ui/icons/Close';
import Dialog from '@material-ui/core/Dialog';
import Paper from '@material-ui/core/Paper';
import PopOutIcon from '@material-ui/icons/Launch';
import Separator from '../atoms/Separator';
import Tooltip from '@material-ui/core/Tooltip';
import ViewerContainer, { componentMap } from '../components/viewers/ViewerContainer';
import { ViewerConfig } from '../components/viewers/Viewer';
import { color, fontsize } from '../Css';
import { stylesheet, classes } from 'typestyle';

const css = stylesheet({
  dialogTitle: {
    color: color.strong,
    fontSize: fontsize.large,
    width: '100%',
  },
  fullscreenCloseBtn: {
    minHeight: 0,
    minWidth: 0,
    padding: 3,
  },
  fullscreenDialog: {
    alignItems: 'center',
    justifyContent: 'center',
    minHeight: '80%',
    minWidth: '80%',
    padding: 20,
  },
  fullscreenViewerContainer: {
    alignItems: 'center',
    boxSizing: 'border-box',
    display: 'flex',
    flexFlow: 'column',
    flexGrow: 1,
    height: '100%',
    justifyContent: 'center',
    margin: 20,
    overflow: 'auto',
    width: '100%',
  },
  plotCard: {
    flexShrink: 0,
    margin: 20,
    minWidth: 250,
    padding: 20,
    width: 'auto',
  },
  plotHeader: {
    display: 'flex',
    justifyContent: 'space-between',
    overflow: 'hidden',
    paddingBottom: 10,
  },
  plotTitle: {
    color: color.strong,
    fontSize: 12,
    fontWeight: 'bold',
  },
  popoutIcon: {
    fontSize: 18,
  },
});

export interface PlotCardProps {
  title: string;
  configs: ViewerConfig[];
  maxDimension?: number;
}

interface PlotCardState {
  fullscreenDialogOpen: boolean;
}

class PlotCard extends React.Component<PlotCardProps, PlotCardState> {
  constructor(props: any) {
    super(props);

    this.state = {
      fullscreenDialogOpen: false,
    };
  }

  public shouldComponentUpdate(nextProps: PlotCardProps, nextState: PlotCardState): boolean {
    return (
      JSON.stringify(nextProps) !== JSON.stringify(this.props) ||
      nextState.fullscreenDialogOpen !== this.state.fullscreenDialogOpen
    );
  }

  public render(): JSX.Element | null {
    const { title, configs, maxDimension, ...otherProps } = this.props;

    if (!configs || !configs.length) {
      return null;
    }

    return (
      <div>
        <Paper {...otherProps} className={classes(css.plotCard, 'plotCard')}>
          <div className={css.plotHeader}>
            <div className={css.plotTitle} title={title}>
              {title}
            </div>
            <div>
              <Button
                onClick={() => this.setState({ fullscreenDialogOpen: true })}
                style={{ padding: 4, minHeight: 0, minWidth: 0 }}
                className='popOutButton'
              >
                <Tooltip title='Pop out'>
                  <PopOutIcon classes={{ root: css.popoutIcon }} />
                </Tooltip>
              </Button>
            </div>
          </div>
          <ViewerContainer configs={configs} maxDimension={maxDimension} />
        </Paper>

        <Dialog
          open={!!this.state.fullscreenDialogOpen}
          classes={{ paper: css.fullscreenDialog }}
          onClose={() => this.setState({ fullscreenDialogOpen: false })}
        >
          <div className={css.dialogTitle}>
            <Button
              onClick={() => this.setState({ fullscreenDialogOpen: false })}
              className={classes(css.fullscreenCloseBtn, 'fullscreenCloseButton')}
            >
              <CloseIcon />
            </Button>
            {componentMap[configs[0].type].prototype.getDisplayName()}
            <Separator />
            <span style={{ color: color.inactive }}>({title})</span>
          </div>
          <div className={css.fullscreenViewerContainer}>
            <ViewerContainer configs={configs} />
          </div>
        </Dialog>
      </div>
    );
  }
}

export default PlotCard;
