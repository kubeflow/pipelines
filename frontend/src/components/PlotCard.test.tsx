/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { shallow } from 'enzyme';
import PlotCard from './PlotCard';
import { ViewerConfig, PlotType } from './viewers/Viewer';

describe('PlotCard', () => {
  it('handles no configs', () => {
    expect(shallow(<PlotCard title='' configs={[]} maxDimension={100} />)).toMatchSnapshot();
  });

  const config: ViewerConfig = { type: PlotType.CONFUSION_MATRIX };

  it('renders on confusion matrix viewer card', () => {
    const tree = shallow(<PlotCard title='test title' configs={[config]} maxDimension={100} />);
    expect(tree).toMatchSnapshot();
  });

  it('pops out a full screen view of the viewer', () => {
    const tree = shallow(<PlotCard title='' configs={[config]} maxDimension={100} />);
    tree.find('.popOutButton').simulate('click');
    expect(tree).toMatchSnapshot();
  });

  it('close button closes full screen dialog', () => {
    const tree = shallow(<PlotCard title='' configs={[config]} maxDimension={100} />);
    tree.find('.popOutButton').simulate('click');
    tree.find('.fullscreenCloseButton').simulate('click');
    expect(tree).toMatchSnapshot();
  });

  it('clicking outside full screen dialog closes it', () => {
    const tree = shallow(<PlotCard title='' configs={[config]} maxDimension={100} />);
    tree.find('.popOutButton').simulate('click');
    tree.find('WithStyles(Dialog)').simulate('close');
    expect(tree).toMatchSnapshot();
  });
});
