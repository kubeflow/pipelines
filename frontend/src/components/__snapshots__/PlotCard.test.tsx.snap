// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlotCard clicking outside full screen dialog closes it 1`] = `
<div>
  <WithStyles(Paper)
    className="plotCard plotCard"
  >
    <div
      className="plotHeader"
    >
      <div
        className="plotTitle"
        title=""
      />
      <div>
        <WithStyles(Button)
          className="popOutButton"
          onClick={[Function]}
          style={
            Object {
              "minHeight": 0,
              "minWidth": 0,
              "padding": 4,
            }
          }
        >
          <WithStyles(Tooltip)
            title="Pop out"
          >
            <pure(LaunchIcon)
              classes={
                Object {
                  "root": "popoutIcon",
                }
              }
            />
          </WithStyles(Tooltip)>
        </WithStyles(Button)>
      </div>
    </div>
    <ViewerContainer
      configs={
        Array [
          Object {
            "type": "confusion_matrix",
          },
        ]
      }
      maxDimension={100}
    />
  </WithStyles(Paper)>
  <WithStyles(Dialog)
    classes={
      Object {
        "paper": "fullscreenDialog",
      }
    }
    onClose={[Function]}
    open={false}
  >
    <div
      className="dialogTitle"
    >
      <WithStyles(Button)
        className="fullscreenCloseBtn fullscreenCloseButton"
        onClick={[Function]}
      >
        <pure(CloseIcon) />
      </WithStyles(Button)>
      Confusion matrix
      <Separator />
      <span
        style={
          Object {
            "color": "#5f6368",
          }
        }
      >
        (
        )
      </span>
    </div>
    <div
      className="fullscreenViewerContainer"
    >
      <ViewerContainer
        configs={
          Array [
            Object {
              "type": "confusion_matrix",
            },
          ]
        }
      />
    </div>
  </WithStyles(Dialog)>
</div>
`;

exports[`PlotCard close button closes full screen dialog 1`] = `
<div>
  <WithStyles(Paper)
    className="plotCard plotCard"
  >
    <div
      className="plotHeader"
    >
      <div
        className="plotTitle"
        title=""
      />
      <div>
        <WithStyles(Button)
          className="popOutButton"
          onClick={[Function]}
          style={
            Object {
              "minHeight": 0,
              "minWidth": 0,
              "padding": 4,
            }
          }
        >
          <WithStyles(Tooltip)
            title="Pop out"
          >
            <pure(LaunchIcon)
              classes={
                Object {
                  "root": "popoutIcon",
                }
              }
            />
          </WithStyles(Tooltip)>
        </WithStyles(Button)>
      </div>
    </div>
    <ViewerContainer
      configs={
        Array [
          Object {
            "type": "confusion_matrix",
          },
        ]
      }
      maxDimension={100}
    />
  </WithStyles(Paper)>
  <WithStyles(Dialog)
    classes={
      Object {
        "paper": "fullscreenDialog",
      }
    }
    onClose={[Function]}
    open={false}
  >
    <div
      className="dialogTitle"
    >
      <WithStyles(Button)
        className="fullscreenCloseBtn fullscreenCloseButton"
        onClick={[Function]}
      >
        <pure(CloseIcon) />
      </WithStyles(Button)>
      Confusion matrix
      <Separator />
      <span
        style={
          Object {
            "color": "#5f6368",
          }
        }
      >
        (
        )
      </span>
    </div>
    <div
      className="fullscreenViewerContainer"
    >
      <ViewerContainer
        configs={
          Array [
            Object {
              "type": "confusion_matrix",
            },
          ]
        }
      />
    </div>
  </WithStyles(Dialog)>
</div>
`;

exports[`PlotCard handles no configs 1`] = `""`;

exports[`PlotCard pops out a full screen view of the viewer 1`] = `
<div>
  <WithStyles(Paper)
    className="plotCard plotCard"
  >
    <div
      className="plotHeader"
    >
      <div
        className="plotTitle"
        title=""
      />
      <div>
        <WithStyles(Button)
          className="popOutButton"
          onClick={[Function]}
          style={
            Object {
              "minHeight": 0,
              "minWidth": 0,
              "padding": 4,
            }
          }
        >
          <WithStyles(Tooltip)
            title="Pop out"
          >
            <pure(LaunchIcon)
              classes={
                Object {
                  "root": "popoutIcon",
                }
              }
            />
          </WithStyles(Tooltip)>
        </WithStyles(Button)>
      </div>
    </div>
    <ViewerContainer
      configs={
        Array [
          Object {
            "type": "confusion_matrix",
          },
        ]
      }
      maxDimension={100}
    />
  </WithStyles(Paper)>
  <WithStyles(Dialog)
    classes={
      Object {
        "paper": "fullscreenDialog",
      }
    }
    onClose={[Function]}
    open={true}
  >
    <div
      className="dialogTitle"
    >
      <WithStyles(Button)
        className="fullscreenCloseBtn fullscreenCloseButton"
        onClick={[Function]}
      >
        <pure(CloseIcon) />
      </WithStyles(Button)>
      Confusion matrix
      <Separator />
      <span
        style={
          Object {
            "color": "#5f6368",
          }
        }
      >
        (
        )
      </span>
    </div>
    <div
      className="fullscreenViewerContainer"
    >
      <ViewerContainer
        configs={
          Array [
            Object {
              "type": "confusion_matrix",
            },
          ]
        }
      />
    </div>
  </WithStyles(Dialog)>
</div>
`;

exports[`PlotCard renders on confusion matrix viewer card 1`] = `
<div>
  <WithStyles(Paper)
    className="plotCard plotCard"
  >
    <div
      className="plotHeader"
    >
      <div
        className="plotTitle"
        title="test title"
      >
        test title
      </div>
      <div>
        <WithStyles(Button)
          className="popOutButton"
          onClick={[Function]}
          style={
            Object {
              "minHeight": 0,
              "minWidth": 0,
              "padding": 4,
            }
          }
        >
          <WithStyles(Tooltip)
            title="Pop out"
          >
            <pure(LaunchIcon)
              classes={
                Object {
                  "root": "popoutIcon",
                }
              }
            />
          </WithStyles(Tooltip)>
        </WithStyles(Button)>
      </div>
    </div>
    <ViewerContainer
      configs={
        Array [
          Object {
            "type": "confusion_matrix",
          },
        ]
      }
      maxDimension={100}
    />
  </WithStyles(Paper)>
  <WithStyles(Dialog)
    classes={
      Object {
        "paper": "fullscreenDialog",
      }
    }
    onClose={[Function]}
    open={false}
  >
    <div
      className="dialogTitle"
    >
      <WithStyles(Button)
        className="fullscreenCloseBtn fullscreenCloseButton"
        onClick={[Function]}
      >
        <pure(CloseIcon) />
      </WithStyles(Button)>
      Confusion matrix
      <Separator />
      <span
        style={
          Object {
            "color": "#5f6368",
          }
        }
      >
        (
        test title
        )
      </span>
    </div>
    <div
      className="fullscreenViewerContainer"
    >
      <ViewerContainer
        configs={
          Array [
            Object {
              "type": "confusion_matrix",
            },
          ]
        }
      />
    </div>
  </WithStyles(Dialog)>
</div>
`;
